### ISSUE TITLE
Feature: Remove printing about "[APIv1] KEEPALIVE /api/v1/events"

### ISSUE THREAD

**Issue by clytras** [mailhog/MailHog#353]
Inside [`api/v1.go`](https://github.com/mailhog/MailHog/blob/0441dd494b03c9255a9b8e90e3458ebb115eacff/vendor/github.com/mailhog/MailHog-Server/api/v1.go#L100) we see that the `keepalive` function uses `stream.Notify("keepalive", []byte{})` to, as stated in the comment, keep connections alive.

```go
 // keepalive sends an empty keep alive message. 
 // 
 // This not only can keep connections alive, but also will detect broken 
 // connections. Without this it is possible for the server to become 
 // unresponsive due to too many open files. 
 func (apiv1 *APIv1) keepalive() { 
 	log.Println("[APIv1] KEEPALIVE /api/v1/events") 
 	stream.Notify("keepalive", []byte{}) 
 } 
```

Which will spam an attached console with many messages like this:

![mailhog spam](https://user-images.githubusercontent.com/6103077/109298593-86cec200-783c-11eb-8cfe-725298ba0268.png)

Can we keep the `stream.Notify("keepalive", []byte{})` which will do the job and ditch the printing `log.Println("[APIv1] KEEPALIVE /api/v1/events")`? Or make an enviroment variable flag option to indicate printing that line or not?

In my developer setup (*and I believe in others too*) I use attached containers to drive every log message to stdout and these event spam messages really bug me up.

**Comment by dbaggott** [comment]
Fyi, related to / arguably duplicate of https://github.com/mailhog/MailHog/issues/56

You can completely disable logging at the risk of losing knowledge of real errors as suggested here:

https://github.com/mailhog/MailHog/issues/56#issuecomment-376463964

**Comment by clytras** [comment]
@dbaggott thanks for the reply.

My request **is NOT** to turn off verbose logging, my request is about killing that useless log line I pointed out. That log line does not offer any kind of useful information. Can you please tell me your opinion on why such a log info might be useful to the developers?

So to clear this out, I want to have logging but I would like to get rid of that useless information about a _hack_ regarding keeping alive the service running.

**Comment by webkp** [comment]
I can only agree @clytras. I found your issue because I was searching for a solution for exactly the same reason. It is annoying that important messages can be overlooked because of the countless ```[APIv1] KEEPALIVE /api/v1/events``` lines.

**Comment by dbaggott** [comment]
@clytras, I was just trying to be helpful in case you didn't see the related issue.

Yes, I imagine once #56 is implemented then there would be support for user configurable log levels and it would be easy to exclude the more verbose debug / info logging...

**Comment by clytras** [comment]
And you did very well @dbaggott, because https://github.com/mailhog/MailHog/issues/56 is related to this _request_ if someone thinks that this line offers some kind of usefull information, they can keep it while debuging with setting the DEBUG logging level if https://github.com/mailhog/MailHog/issues/56 is picked and there will be a logging system with configurable levels. But again, I think that this log line needs to be deleted as it spams not only the terminal buffer and screen in case of `stdout` logging, but also log files with countless of useless information.

**Comment by berkeleybross** [comment]
Im not entirely confident in GO, but I can see that there is already a debug function in https://github.com/mailhog/MailHog/blob/0441dd494b03c9255a9b8e90e3458ebb115eacff/vendor/gopkg.in/mgo.v2/log.go#L105

Can we use that instead? Happy to make a PR with that change

**Comment by berkeleybross** [comment]
Apologies, should have looked closer at the code :) Looks like that line is actually from the mongo package used by MailHog.

There is an import statement to https://github.com/ian-kent/go-log at the top of the api file, which looks like it supports debug as well - that would be a more suitable fix :) 

Any objections for me to raise a PR?

**Comment by clytras** [comment]
For anyone interested on getting rid of that annoying spam message, I've forked this and created my own Docker Hub image [`lytrax/mailhog`](https://hub.docker.com/r/lytrax/mailhog) with tha `latest` tag like `lytrax/mailhog:latest`.

I've tested it with Docker and Laravel Sail with different binding ports even and it works pretty well so far without spaming the output console every minute.

You can even create your own images using the [`Dockerfile`](https://github.com/clytras/MailHog/blob/master/Dockerfile) in my forked repo which `go get` will download the source from my repo.

**Comment by Sam-Lane** [comment]
I agree with @clytras this log message is of no benefit and just creates noise in my logs. I would be really happy to see it removed/silenced.

**Comment by macau23** [comment]
I'd like to keep this log message, but change the log level to make it invisible by default.

**Comment by smjet** [comment]
Any ETA on this? It's been almost a year.

**Comment by olevekavitocz** [comment]
2 year without proper answer? :-D Cmon guyz help us with our paper trail test server situation :-)

**Comment by bradjones1** [comment]
> 2 year without proper answer? :-D Cmon guyz help us with our paper trail test server situation :-)

This is not helpful.

This is an open-source project, and for those who are complaining and enquiring about when this might change... you are more than welcome to propose a PR or hire a developer (one of the project maintainers, maybe?) to do so, for you. Nobody in open-source owes you anything if you aren't paying for it.

**Comment by fnep** [comment]
I'm currently using a custom entry point as a workaround. Not happy with it, but it works for me.

```yaml
  mailserver:
    image: mailhog/mailhog
    entrypoint: ["/bin/sh", "-c", "MailHog | grep -v KEEPALIVE"]
```  

_(edited to avoid people copying the wrong version)_

**Comment by olevekavitocz** [comment]
> > 2 year without proper answer? :-D Cmon guyz help us with our paper trail test server situation :-)
> 
> This is not helpful.
> 
> This is an open-source project, and for those who are complaining and enquiring about when this might change... you are more than welcome to propose a PR or hire a developer (one of the project maintainers, maybe?) to do so, for you. Nobody in open-source owes you anything if you aren't paying for it.

And the winner of the competition for the most boring person is... @bradjones1! Gratz, buddy.

Of course we can write a PR, of course we can hire a programmer to do it for us, or we can write it ourselves (which I already did in the fork version), but that doesn't address my comment, which is that for two years this thread has been open and unanswered. We have been dealing with this problem in the company for a long time and I periodically look at possible solutions here in the abandoned topic.

I have not attacked the authors in any way and have tried to nudge them towards a solution in a very friendly way.

Thanks @fnep, we use a similar solution, and like you we are not thrilled with it.

**Comment by mfn** [comment]
> I'm currently using a custom entry point as a workaround. Not happy with it, but it works for me.

I tried this one too, but I still got the output… delayed; like it was buffered. I didn't dug further, but is the output maybe sent to stderr? 🤔

I also tried this in docker compose, but I must have been doing something wrong, I still got the output; and I didn't favour this one because I still want to get regular logs:
```
    logging:
      driver: none
```

@olevekavitocz can you share your solution?

**Comment by olevekavitocz** [comment]
We use the almost same design as @fnep wrote. The application is wrapped in a docker and we filter the output from mailhog using `grep -v "[APIv1] KEEPALIVE"` (removes lines from the output that goes to the log) for a custom entrypoint. The filtered content is then saved to syslog.

Alternatively (older projects) we use a "custom" (fork) mailhog, where only the logging line is commented out (as @clytras originally wrote) - which I probably don't recommend.

There is also the possibility to filter the log as such at the operating system level (I have no experience with this, but similar solutions are used by companies we work with) e.g. https://github.com/syslog-ng/syslog-ng

**Comment by fnep** [comment]
@olevekavitocz 

I had just assembled this fresh for the comment and maybe did it too narrow. 

In the real config we simply grep for KEEPALIVE like this (what is even worse):

```yaml
  mailserver:
    image: mailhog/mailhog
    entrypoint: ["/bin/sh", "-c", "MailHog | grep -v KEEPALIVE"]
```

**Comment by mfn** [comment]
Can confirm that this version finally worked; the previous grep for some reason didn't:
`entrypoint: ["/bin/sh", "-c", "MailHog | grep -v KEEPALIVE"]`

I also wanted to disable buffering as I've noticed it (`--line-buffering`), but the version in the container doesn't feature it (or I missed it).

**Comment by egor83** [comment]
I wanted to disable logging altogether, so I combined @fnep's solution and @talawahtech's solution in #56 (thanks for these!) a bit, editing docker-compose.yml as follows:

```
mailhog:
  ...
  entrypoint: ["/bin/sh", "-c", "MailHog &>/dev/null"]
```

**Comment by ap0ca** [comment]
This works for me, it removes KEEPALIVE logs but keeps the rest of the logs. I switched from `grep` to `awk`:

`
  entrypoint: ["/bin/sh", "-c", "MailHog | awk '!/KEEPALIVE/'"]
`
